sklearn/neighbors/kd_tree.pyx

```diff
@@ -15,34 +15,19 @@ VALID_METRICS = ['EuclideanDistance', 'ManhattanDistance',
                  'ChebyshevDistance', 'MinkowskiDistance']
 
 
-class SubstituteDoc(type):
-    def __init__(cls, clsname, superclasses, attributedict):
-        for d in dir(cls):
-            #Exclude special methods and non functions
-            if d.startswith("__"):
-                continue
-            dc = getattr(cls, d)
-            if not callable(dc):
-                continue
-            #Copy the function, this is necessary to avoid modifying the docstrings of the superclass
-            g = types.FunctionType(dc.__code__, dc.__globals__, name=dc.__name__,
-                           argdefs=dc.__defaults__,
-                           closure=dc.__closure__)
-            g = functools.update_wrapper(g, dc)
-            g.__kwdefaults__ = dc.__kwdefaults__
-
-            g.__doc__ = dc.__doc__.format(**DOC_DICT)
-            setattr(cls,d,g)
+
 
 # Inherit KDTree from BinaryTree
 include "binary_tree.pxi"
-import types
-import functools
+
 
 cdef class KDTree(BinaryTree):
-    __metaclass__ = SubstituteDoc
+    # __metaclass__ = SubstituteDoc
     __doc__ = CLASS_DOC.format(**DOC_DICT)
-    pass
+
+    def __init__(self, data, leaf_size=40, metric='minkowski', **kwargs):
+        super(KDTree,self).__init__(data, leaf_size, metric, **kwargs)
+        self.substituteDoc(DOC_DICT)
```

Comment:
@jnothman I added the method in BinaryTree and called it in its children but it still doesn't seem to affect the final documentation:
https://15271-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.neighbors.KDTree.html#sklearn.neighbors.KDTree.query
Have I missed something? 
If you point me to the right resources I could try to understand why this isn't considered while producing the docs.

Reply:
Yes, it does seem to affect the final documentation. Previously it was rendering `{BinaryTree}`. [Now](https://15327-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.neighbors.KDTree.html) it renders `BinaryTree`. it has made a substitution, but the wrong one....

Reply:
Oh, now I'm confused about that statement of mine.... I now see you're doing the substitution in `__init__`. No, surely that won't work, as the docs are compiled from the classes, not from initialised instances. You'll need to use a metaclass or in the lines after the class is defined: